vncviewer/UserDialog.cxx

```diff
@@ -62,34 +63,54 @@ static void button_cb(Fl_Widget *widget, void *val) {
   widget->window()->hide();
 }
 
+static void keepPassword(Fl_Widget *widget, void *data) {
+  UserDialog *dialog = (UserDialog*) data;
+  Fl_Check_Button* flcb = (Fl_Check_Button*)widget;
+  dialog->setKeepPassword(flcb->value());
+}
+
 UserDialog::UserDialog()
 {
+  user_ = NULL;
+  password_ = NULL;
+  keepPasswd_ = true;
 }
 
 UserDialog::~UserDialog()
 {
+  resetPassword();
+}
+
+void UserDialog::resetPassword()
+{
+  if(user_) { free(user_); user_ = NULL; }
+  if(password_) { free(password_); password_ = NULL; }
 }
 
 void UserDialog::getUserPasswd(bool secure, char** user, char** password)
 {
   CharArray passwordFileStr(passwordFile.getData());
 
   assert(password);
-  char *envUsername = getenv("VNC_USERNAME");
-  char *envPassword = getenv("VNC_PASSWORD");
 
-  if(user && envUsername && envPassword) {
-    *user = strdup(envUsername);
-    *password = strdup(envPassword);
+  if(password_) {
+    if(user_) *user = strdup(user_);
```

Comment:
This will crash. You need to check if `user` is valid.

(same thing a few other places)